constraint_generation/src/execute.rs

```diff
@@ -207,7 +223,40 @@ fn execute_statement(
                     let symbol = AExpr::Signal { symbol: constrained.left };
                     let expr = AExpr::sub(&symbol, &constrained.right, &p);
                     let ctr = AExpr::transform_expression_to_constraint_form(expr, &p).unwrap();
-                    node.add_constraint(ctr);
+                    if constrained.custom_gate_name.is_none() {
+                        node.add_constraint(ctr);
+                    } else {
+                        // From a previous semantic analysis we know that in this case we must
+                        // have that constrained.right is an AExpr::Signal, so we can safely unwrap
+                        // the name of the signal in the right hand side of the expression.
+                        debug_assert!(matches!(symbol, AExpr::Signal {..}));
+                        debug_assert!(matches!(constrained.right, AExpr::Signal {..}));
+                        if let AExpr::Signal { symbol: left } = symbol {
+                            if let AExpr::Signal { symbol: right } = constrained.right {
+                                let custom_gate_name = constrained.custom_gate_name.unwrap();
+                                fn reorder(
+                                    left: String,
+                                    right: String,
+                                    custom_gate_name: &String
+                                ) -> (String, String) {
+                                    if left.starts_with(custom_gate_name) {
+                                        (left, right)
+                                    } else {
+                                        debug_assert!(right.starts_with(custom_gate_name));
+                                        (right, left)
+                                    }
+                                }
+
+                                // Assignment of the form left <== right
+                                let (inner, outer) = reorder(left, right, &custom_gate_name);
```

Comment:
We may have a substitution like signal ==> cg.input or cg.output ==> signal, but we want to store that as a pair of strings where the first element of the pair is the signal in the custom gate.